README.md
# array-slayer

> 🚨🚨 This project is a work in progress! Issues and pull requests are encouraged. 🚨🚨

## About
With the array-slayer utility library you have:

- smooth experice of functional programming tackling daily problems with arrays.
- faster development speed by letting array-slayer solve usual problems for you.
- ability to chain methods or import a single method as a seperate module.
- much more readable & consice code for reviewing.
- lightweight import: only about ~16kbs (full-featured bundle)
- immutability by nature (methods return a new array)

## Example usage
```js
import ArraySlayer from "arraySlayer";

const array = [
    [47,25,1,{ a: "Foo2", b: "Bar1", c: { d: "ct", e: { f: 9 }, g: 1 } }],
    [47,29,2,{ a: "Foo1", b: "Bar2", c: { d: "bg", e: { f: 5 }, g: 4 } }],
    [47,25,3,{ a: "Foo3", b: "Bar3", c: { d: "ce", e: { f: 7 }, g: 3 } }],
    [47,38,4,{ a: "Foo4", b: "Bar4", c: { d: "zd", e: { f: 1 }, g: 4 } }],
];

const result = ArraySlayer(array)
                .column(3, "c.e.f")
                .concat([17, 2, 1, 18])
                .replaceByValue(1, 100)
                .unique()
                .sort({ order: "ascending", sortType: "QuickSort"})
                .value;

console.log(result); //-> [2, 5, 7, 9, 17, 18, 100]
```

of if you need a single method from arraySlayer, you can import it as a seperate module:

```js
import { sort } from "array-slayer/sort.js";

const result = sort(array, { sortType: "MergeSort", sortBy: item => item[3].c.d });

console.log(result); // returns sorted array based on the selected keys
```

## Quick Links
1. [at](#at)
1. [booleans](#booleans)
    - [AND](#and)
    - [AND_ALL](#andall)
    - [AND_OR](#andor)
    - [OR](#or)
    - [OR_ALL](#orall)
    - [OR_AND](#orand)
    - [XOR_ALL](#xorall)
1. [chunk](#chunk)
1. [clear](#clear)
1. [clearByIndexes](#clearbyindexes)
1. [clearByValues](#clearbyvalues)
1. [column](#column)
1. [count](#count)
1. [deleteByIndexes](#deletebyindexes)
1. [deleteByValues](#deletebyvalues)
1. [difference](#difference)
1. [findAllOccurences](#findAllOccurences)
1. [findDuplicates](#findDuplicates)
1. [get](#get)
    - [getTruthyValues](#gettruthyvalues)
    - [getFalsyValues](#getfalsyvalues)
    - [getAllKeys](#getallkeys)
1. [has](#has)
    - [has](#has)
    - [hasAll](#hasall)
    - [hasAllTypes](#hasalltypes)
    - [hasArray](#hasarray)
    - [hasBoolean](#hasboolean)
    - [hasDuplicates](#hasduplicates)
    - [hasEmptyArr](#hasemptyarr)
    - [hasEmptyObj](#hasemptyobj)
    - [hasFalse](#hasfalse)
    - [hasFalsy](#hasfalsy)
    - [hasNaN](#hasnan)
    - [hasNull](#hasnull)
    - [hasNullish](#hasnullish)
    - [hasNumber](#hasnumber)
    - [hasObject](#hasobject)
    - [hasString](#hasstring)
    - [hasTrue](#hastrue)
    - [hasTruthy](#hastruthy)
    - [hasUndefined](#hasundefined)
1. [interpolate](#interpolate)
1. [intersection](#intersection)
1. [is](#is)
    - [isEmpty](#isempty)
    - [isEqual](#isequal)
    - [isUnique](#isunique)
1. [math](#math)
    - [max](#max)
    - [maxBy](#maxby)
    - [maxOf](#maxof)
    - [mean](#mean)
    - [meanOf](#meanof)
    - [min](#min)
    - [minBy](#minby)
    - [minOf](#minof)
    - [product](#product)
    - [productOf](#productof)
    - [sum](#sum)
    - [sumOf](#sumof)
1. [matrix](#matrix)
    - [dotProduct](#dotproduct)
    - [crossProduct](#crossproduct)
1. [random](#random)
1. [range](#range)
1. [readOnly](#readOnly)
1. [replace](#replace)
    - [replaceByIndex](#replacebyindex)
    - [replaceByIndexes](#replacebyindexes)
    - [replaceByValue](#replacebyvalue)
    - [replaceByValues](#replacebyvalues)
1. [rotate](#rotate)
1. [shuffle](#shuffle)
1. [sort](#sort)
1. [swap](#swap)
    - [swapByIndexes](#swapbyindexes)
    - [swapByValues](#swapbyvalues)
1. [symmetricDifference](#symmetricDifference)
1. [take](#take)
1. [toJSONObject](#toJSONObject)
1. [types](#types)
1. [union](#union)
1. [unique](#unique)
    - [unique](#unique)
    - [uniqueBy](#uniqueby)
1. [x](#x)
    - [xpop](#xpop)
    - [xpush](#xpush)
    - [xshift](#xshift)
    - [xunshift](#xunshift)
    - [xsplice](#xsplice)
1. [zip](#zip)
    - [zip](#zip)
    - [unzip](#unzip)

**[⬆ back to top](#quick-links)**

## Specific Chain Methods
- [Native JS Methods](#native-js-methods)
- [deepCopy](#deepcopy)
- [lastElement](#lastelement)
- [length](#length)
- [mutate](#mutate)
- [next](#next)
- [value](#value)

## Advanced Configuration
- [configuring deepCopy](#configuring-deepcopy)
- [configuring serialization](#configuring-serialization)

**[⬆ back to top](#quick-links)**


## **at**

returns the value from the given array index. (argument can also be a negative number corresponding to counting from end of the array)

```js
import Arr from "array-slayer";

const numbers = [1, 2, 3, 4, 5];

const secondItem = Arr(numbers).at(1);       // -> 2
const lastItem = Arr(numbers).at(-1);        // -> 5
const nextToLastItem = Arr(numbers).at(-2);  // -> 4
```

**[⬆ back to top](#quick-links)**

## booleans

#### **AND**

can check if every single item in array is equal to the given constant

```js
import Bools from "array-slayer/booleans.js";

const flag = true;
const array = [true, true, true, true];

// Bools([a, b, c]).AND(flag) => (a === flag) && (b === flag) && (c === flag)
const result = Bools(array).AND(flag);
console.log(result); // -> true
```

can also check if a filter function passes on every single item in an array

```js
import Bools from "array-slayer/booleans.js";

const id = 18;
const array = [-1, 5, 2, 4];

// Bools([a, b, c]).AND(item, fn) => (fn(a, item)) && (fn(b, item)) && (fn(c, item))
const result = Bools(array).AND(id, item => item < id);
console.log(result); // -> true
```

can also check if two arrays of the same length, have the same items in the same order

```js
import Bools from "array-slayer/booleans.js";

const array1 = [-1, 5, 2, true];
const array2 = [-1, 5, 2, false];

// Bools([a, b, c]).AND([e, f, g]) => (a === e) && (b === f) && (c === g)
const result = Bools(array1).AND(array2);
console.log(result); // -> false
```

can also check if a function passes on each elements of the two arrays of the same length, respectively

```js
import Bools from "array-slayer/booleans.js";

const array1 = [-1, 5, 2, 8];
const array2 = [-3, 4, 1, 7];

// Bools([a, b, c]).AND([e, f, g], fn) => (fn(a,e)) && (fn(b,f)) && (fn(c,g))
const result = Bools(array1).AND(array2, (a,b) => a > b);
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **AND_ALL**

ANDs all elements of the given array (& optional argument)

```js
import Bools from "array-slayer/booleans.js";

let flag = true;
const array = [1, true, {}, "name"];

// Bools([a, b, c]).AND_ALL(bool) => (a && bool) && (b && bool) && (c && bool)
const result = Bools(array).AND_ALL(flag);
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **AND_OR**

ORs all elements of the given array & ANDs the result with an optional argument

```js
import Bools from "array-slayer/booleans.js";

const flag = false;
const array = [1, 2, 3, true, {}];

// Bools([a, b, c]).AND_OR(bool) => (a && bool) || (b && bool) || (c && bool)
const result = Bools(array).AND_OR();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **OR**

can check if at least one item in array is equal to the given constant

```js
import Bools from "array-slayer/booleans.js";

const flag = true;
const array = [false, false, false, true];

// Bools([a, b, c]).OR(flag) => (a === flag) || (b === flag) || (c === flag)
const result = Bools(array).OR(flag);
console.log(result); // -> true
```

can also check if a filter function passes on at least one item in an array

```js
import Bools from "array-slayer/booleans.js";

const id = 18;
const array = [40, 50, 2, 60];

// Bools([a, b, c]).OR(item, fn) => (fn(a, item)) || (fn(b, item)) || (fn(c, item))
const result = Bools(array).OR(id, item => item < id);
console.log(result); // -> true
```

can also check if two arrays of the same length, have at least one common item with the same index

```js
import Bools from "array-slayer/booleans.js";
// or 'import Bools from "array-slayer";' alternatively

const array1 = [-1, 5, 2, true];
const array2 = [-1, 1, 7, false];

// Bools([a, b, c]).OR([e, f, g]) => (a === e) || (b === f) || (c === g)
const result = Bools(array1).OR(array2);
console.log(result); // -> true
```

can also check if a function passes at least one of the respective elements of the two arrays of the same length

```js
import Bools from "array-slayer/booleans.js";

const array1 = [70, 1, 2, 8];
const array2 = [-3, 4, 1, 7];

// Bools([a, b, c]).OR([e, f, g], fn) => (fn(a,e)) || (fn(b,f)) || (fn(c,g))
const result = Bools(array1).OR(array2, (a,b) => a < b);
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **OR_ALL**

ORs all elements of the given array (& optional argument)

```js
import Bools from "array-slayer/booleans.js";

const array = [1, 2, 3, 0];

// Bools([a, b, c]).OR_ALL(bool) => (a || bool) && (b || bool) && (c || bool)
const result = Bools(array).OR_ALL();
console.log(result); // -> false
```

**[⬆ back to top](#quick-links)**

#### **OR_AND**

ANDs all elements of the given array & ORs the result with  an optional argument

```js
import Bools from "array-slayer/booleans.js";

const flag = true;
const array = [1, 2, 3, 0];

// Bools([a, b, c]).OR_AND(bool) => (a || bool) && (b || bool) && (c || bool)
const result = Bools(array).OR_AND(flag);
console.log(result); // -> true
```

#### **XOR_ALL**

XORs all elements of the given array (& optional argument)

```js
import Bools from "array-slayer/booleans.js";

const array = [1, 2, 3, 0];

const result = Bools(array).XOR_ALL();
console.log(result); // -> 0
```

**[⬆ back to top](#quick-links)**

## chunk

```js
import { chunk } from "array-slayer/chunk.js";

const array = [1,2,9,7,8,6,2,6,2,7];
const result = chunk(3, array);
console.log(result); // -> [ [ 1, 2, 9 ], [ 7, 8, 6 ], [ 2, 6, 2 ], [ 7 ] ]
```

or

```js
import ArraySlayer from "array-slayer";

const array = [1,2,9,7,8,6,2,6,2];
const result = ArraySlayer(array).chunk(3).value;
console.log(result); // -> [ [ 1, 2, 9 ], [ 7, 8, 6 ], [ 2, 6, 2 ], [ 7 ] ]
```

**[⬆ back to top](#quick-links)**

## clear

empties an array (mutates the original array, keeping its reference)

```js
import { clear } from "array-slayer/clear.js";

const array = [1,2,9,7,8,6,2,6,2];
const result = clear(array);
console.log(result); // -> []
```

or

```js
import ArraySlayer from "array-slayer";

const array = [1,2,9,7,8,6,2,6,2];
const result = ArraySlayer(array).clear().value;
console.log(result); // -> []
```

**[⬆ back to top](#quick-links)**

## clearByIndexes

replaces the values of the given indexes by `undefined` (keeps the array length)

```js
import { clearByIndexes } from "array-slayer/clearByIndexes.js";

const array = [1,2,9,7,8,6,2,6,2];
const result = clearByIndexes(array, 0, 2, 3);
console.log(result); // -> [undefined,2,undefined,undefined,8,6,2,6,2];
```

or

```js
import ArraySlayer from "array-slayer";

const array = [1,2,9,7,8,6,2,6,2];
const result = ArraySlayer(array).clearByIndexes(0, 2, 3).value;
console.log(result); // -> [undefined,2,undefined,undefined,8,6,2,6,2];
```

**[⬆ back to top](#quick-links)**

## clearByValues

replaces the given values of the given array by `undefined` (keeps the array length)

```js
import { clearByValues } from "array-slayer/clearByValues.js";

const array = [1,2,9,7,8,6,2,6,2];
const result = clearByValues(array, 1, 6, 8);
console.log(result); // -> [undefined,2,9,7,undefined,undefined,2,undefined,2];
```

clearByValues can also compare objects, but it is more costly than clearByIndexes

```js
import ArraySlayer from "array-slayer";

const array = [4, {id: 12}, "str", {c: [8, {d: "name"}] }, 24];
const result = ArraySlayer(array).clearByValues("str", { c: [8, {d: "name"}] }).value;
console.log(result); // -> [4, {id: 12}, undefined, undefined, 24];
```

**[⬆ back to top](#quick-links)**

## column

returns the selected column from a 2D array or array of objects

```js
import { column } from "array-slayer/column.js";

const array = [
    [1,2,3],
    [4,5,6],
    [7,8,9],
];
const result = column(array, "1");
console.log(result); // -> [2, 5, 8];
```

or

```js
import ArraySlayer from "array-slayer";

const array = [
    [4, {id: 1}, "baz", {c: [8, {d: "foo"}], f: { e: 5 } }, 24],
    [4, {id: 2}, "foo", {c: [8, {d: "bar"}], f: { e: 6 } }, 24],
    [4, {id: 3}, "bar", {c: [8, {d: "baz"}], f: { e: 7 } }, 24]
];
const result = ArraySlayer(array).column(3, "f.e").value;
console.log(result); // -> [5, 6, 7];
```

**[⬆ back to top](#quick-links)**

## count

returns how many times an item is repeated in an array

```js
import { count } from "array-slayer/count.js";

const array = [
    [1,2,3],
    [4,5,6],
    [7,8,9],
];
const result = count([4,5,6], array);
console.log(result); // -> 1
```

or

```js
import ArraySlayer from "array-slayer";

const array = [1,2,1,5,1,8];
const result = ArraySlayer(array).count(1);
console.log(result); // -> 3
```

**[⬆ back to top](#quick-links)**

## deleteByIndexes

removes array items by the given array indexes (changes the array length)

```js
import { deleteByIndexes } from "array-slayer/deleteByIndexes.js";

const array = [5,6,8,4,1];
const result = deleteByIndexes(array, 0, 2, 3);
console.log(result); // -> [6,1]
```

or

```js
import ArraySlayer from "array-slayer";

const array = [5,6,8,4,1];
const result = ArraySlayer(array).deleteByIndexes(0, 2, 3).value;
console.log(result); // -> [6,1]
```

**[⬆ back to top](#quick-links)**

## deleteByValues

removes array items by the given array values (changes the array length)

```js
import { deleteByValues } from "array-slayer/deleteByValues.js";

const array = [5,6,8,4,1];
const result = deleteByValues(array, 5, 8, 4);
console.log(result); // -> [6,1]
```

or

```js
import ArraySlayer from "array-slayer";

const array = [5,6,8,4,1];
const result = ArraySlayer(array).deleteByValues(5, 8, 4).value;
console.log(result); // -> [6,1]
```

**[⬆ back to top](#quick-links)**

## difference

returns an array including the first array's values excluding the second array's values

```js
import { difference } from "array-slayer/difference.js";

const array1 = [5,6,8,4,1];
const array2 = [7,8,9,5,3];
const result = difference(array1, array2);
console.log(result); // -> [6, 4, 1]
```

or

```js
import ArraySlayer from "array-slayer";

const array1 = [5,6,8,4,1];
const array2 = [7,8,9,5,3];
const result = ArraySlayer(array1).difference(array2).value;
console.log(result); // -> [6, 4, 1]
```

**[⬆ back to top](#quick-links)**

## findAllOccurences

returns the indexes of all occurences of the selected item in array (by value)

```js
import { findAllOccurences } from "array-slayer/findAllOccurences.js";

const array = [5,6,5,5,1,7,8,5];
const result = findAllOccurences(5);
console.log(result); // -> [0, 2, 3, 7]
```

or

```js
import ArraySlayer from "array-slayer";

const array = [5,6,5,5,1,7,8,5];
const result = ArraySlayer(array).findAllOccurences(5).value;
console.log(result); // -> [0, 2, 3, 7]
```

**[⬆ back to top](#quick-links)**

## findDuplicates

returns the duplicate values in array

```js
import { findDuplicates } from "array-slayer/findDuplicates.js";

const array = [5,6,5,5,1,7,8,5];
const result = findDuplicates(array);
console.log(result); // -> [5]
```

or

```js
import ArraySlayer from "array-slayer";

const array = [5,6,5,5,1,7,8,5];
const result = ArraySlayer(array).findDuplicates(array).value;
console.log(result); // -> [5]
```

**[⬆ back to top](#quick-links)**

## get

#### **getTruthyValues**

```js
import { getTruthyValues } from "array-slayer/get.js";

const array = [4, true, false, 0, 1, "", "name", {}, [], function(){}];
const result = getTruthyValues(array);
console.log(result); // -> [4, true, 1, "", "name", {}, [], f()];
```

or

```js
import ArraySlayer from "array-slayer";

const array = [4, true, false, 0, 1, "", "name", {}, [], function(){}];
const result = ArraySlayer(array).getTruthyValues(array).value;
console.log(result); // -> [4, true, 1, "", "name", {}, [], f()];
```

**[⬆ back to top](#quick-links)**

#### **getFalsyValues**

```js
import { getFalsyValues } from "array-slayer/get.js";

const array = [4, true, false, 0, 1, "", "name", {}, [], function(){}];
const result = getFalsyValues(array);
console.log(result); // -> [false, 0];
```

or

```js
import ArraySlayer from "array-slayer";

const array = [4, true, false, 0, 1, "", "name", {}, [], function(){}];
const result = ArraySlayer(array).getFalsyValues(array).value;
console.log(result); // -> [false, 0];
```

**[⬆ back to top](#quick-links)**

#### **getAllKeys**

```js
import { getAllKeys } from "array-slayer/get.js";

const array = [{id: 12, name: "foo", lastName: "bar"}];
const result = getAllKeys(array[0]);
console.log(result); // -> ["id", "name", "foo", "lastName"];
```

or

```js
import ArraySlayer from "array-slayer";

const array = [{id: 12, name: "foo", lastName: "bar"}];
const result = ArraySlayer(array).getAllKeys(array[0]).value;
console.log(result); // -> ["id", "name", "foo", "lastName"];
```

**[⬆ back to top](#quick-links)**

## has

#### **has**

```js
import { has } from "array-slayer/has.js";

const array = [1,2,3,4,5];
const result = has(array, 3);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [{id: 12, name: "foo", lastName: "bar"}];
const result = ArraySlayer(array).has({id: 12, name: "foo", lastName: "bar"});
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasAll**

```js
import { hasAll } from "array-slayer/has.js";

const array = [1,2,3,4,5];
const result = hasAll(array, [2,4,5]);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [1,2,3,4,5];
const result = ArraySlayer(array).hasAll([2,4,5]);
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasAllTypes**

```js
import { hasAllTypes } from "array-slayer/has.js";

const array = ["foo",false,3,4,5];
const result = hasAllTypes(array, ["string", "boolean", "number"]);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = ["foo",false,3,4,5];
const result = ArraySlayer(array).hasAllTypes(["string", "boolean", "number"]);
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasArray**

```js
import { hasArray } from "array-slayer/has.js";

const array = ["foo", false, {id: 2}, 4, [1, {name: "foo"}],5];
const result = hasArray(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [[],false,3,4,5];
const result = ArraySlayer(array).hasArray();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasBoolean**

```js
import { hasBoolean } from "array-slayer/has.js";

const array = ["foo", false, {id: 2}, 4, [1, {name: "foo"}],5];
const result = hasBoolean(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [[],false,3,4,5];
const result = ArraySlayer(array).hasBoolean();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasDuplicates**

returns true if array has duplicate values

```js
import { hasDuplicates } from "array-slayer/has.js";

const array = [[],false,3,3,3];
const result = hasDuplicates(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [{id: 12},{id: 12},3,4,5];
const result = ArraySlayer(array).hasDuplicates();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasEmptyArr**

returns true if array includes an empty array ([]) // depth = 1

```js
import { hasEmptyArr } from "array-slayer/has.js";

const array = [[],false,3,3,3];
const result = hasEmptyArr(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [{id: 12},[18],3,4,5];
const result = ArraySlayer(array).hasEmptyArr();
console.log(result); // -> false
```

**[⬆ back to top](#quick-links)**

#### **hasEmptyObj**

returns true if array includes an empty object ({}) // depth = 1

```js
import { hasEmptyObj } from "array-slayer/has.js";

const array = [[],{},3,3,3];
const result = hasEmptyObj(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [{id: 12},[18],3,4,5];
const result = ArraySlayer(array).hasEmptyObj();
console.log(result); // -> false
```

**[⬆ back to top](#quick-links)**

#### **hasFalse**

returns true if array includes `false` boolean

```js
import { hasFalse } from "array-slayer/has.js";

const array = [[],{},3,false,3];
const result = hasFalse(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [[],{},0,null,undefined];
const result = ArraySlayer(array).hasFalse();
console.log(result); // -> false
```

**[⬆ back to top](#quick-links)**

#### **hasFalsy**

returns true if array includes a falsy value (`0`, `null`, `undefined`, `false`, ...)

```js
import { hasFalsy } from "array-slayer/has.js";

const array = [[],{},3,0,3];
const result = hasFalsy(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [[],{},1,null,5];
const result = ArraySlayer(array).hasFalsy();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasNaN**

returns true if array includes a `NaN`

```js
import { hasNaN } from "array-slayer/has.js";

const array = [[],{},0, null, undefined, false];
const result = hasNaN(array);
console.log(result); // -> false
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,1,null,5];
const result = ArraySlayer(array).hasNaN();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasNull**

returns true if array includes `Null`

```js
import { hasNull } from "array-slayer/has.js";

const array = [[],{},0, NaN, undefined, false];
const result = hasNull(array);
console.log(result); // -> false
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,1,null,5];
const result = ArraySlayer(array).hasNull();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasNullish**

returns true if array includes `null` or `undefined`

```js
import { hasNullish } from "array-slayer/has.js";

const array = [[],{},0, NaN, undefined, false];
const result = hasNullish(array);
console.log(result); // -> false
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,1,undefined,5];
const result = ArraySlayer(array).hasNullish();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasNumber**

returns true if array includes an item of type `"number"`

```js
import { hasNumber } from "array-slayer/has.js";

const array = [[],{},function(){}, null, undefined, false];
const result = hasNumber(array);
console.log(result); // -> false
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,1,undefined,5];
const result = ArraySlayer(array).hasNumber();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasObject**

returns true if array includes an Object

```js
import { hasObject } from "array-slayer/has.js";

const array = [1, 5, null, undefined, false];
const result = hasObject(array);
console.log(result); // -> false
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,1,{id: "foo"},5];
const result = ArraySlayer(array).hasObject();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasString**

returns true if array includes an item of type `"string"`

```js
import { hasString } from "array-slayer/has.js";

const array = [1, 5, null, {id: "foo"}, false];
const result = hasString(array);
console.log(result); // -> false
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,1,"foo",5];
const result = ArraySlayer(array).hasString();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasTrue**

returns true if array includes a `true` boolean

```js
import { hasTrue } from "array-slayer/has.js";

const array = [1, 5, null, undefined, false];
const result = hasTrue(array);
console.log(result); // -> false
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,1,true,5];
const result = ArraySlayer(array).hasTrue();
console.log(result); // -> true
```

**[⬆ back to top](#quick-links)**

#### **hasTruthy**

returns true if array includes a truthy value

```js
import { hasTruthy } from "array-slayer/has.js";

const array = [1, 5, null, undefined, false];
const result = hasTruthy(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,null,undefined,false];
const result = ArraySlayer(array).hasTruthy();
console.log(result); // -> false
```

**[⬆ back to top](#quick-links)**

#### **hasUndefined**

returns true if array includes `undefined`

```js
import { hasUndefined } from "array-slayer/has.js";

const array = [1, 5, null, undefined, false];
const result = hasUndefined(array);
console.log(result); // -> true
```

or

```js
import ArraySlayer from "array-slayer";

const array = [0,NaN,null,"", [], {},false];
const result = ArraySlayer(array).hasUndefined();
console.log(result); // -> false
```

**[⬆ back to top](#quick-links)**

## interpolate

Linear Interpolation

A -> a

B -> x = ?

C -> c

array1 = [A, B, C]

array2= [a, "x", c]

x = interpolate(array1, array2);

initial value for `x` can be set to either `"x"`, `"X"`, `"?"`, `null`, `false` or `undefined`

```js
import { interpolate } from "array-slayer/interpolate.js";

const array1 = [1,2,3];
const array2 = [2,"?",6];
const result = interpolate(array1, array2);
console.log(result); // -> 4
```

or

```js
import ArraySlayer from "array-slayer";


CHANGELOG.md
# array-slayer changes

## \[0.2.3] - 2021-07-08

added `clean` & `prebuild` commands to package.json

## \[0.2.2] - 2021-07-07

- package was tested locally for publish.
- removed index.js from /fp directory & build.
- docs: fixed import names

## \[0.2.1] - 2021-07-06

added `esbuild` for bundling

## \[0.2.0] - 2021-07-04

added seperatd methods as modules

## \[0.1.0] - 2021-07-01

First public release.


-----------------------------------

### TODO:
- adding more methods: `paginate`, `debounce`, `throttle`
- cleaning `example.js`
- adding typescript?
- adding husky, lint-staged & commit-lint
- writing tests
- future collections: 'Obj', 'Num', 'Matrix', 'Str', 'Date'

lib/examples.js
import Arr from "./src/arraySlayer.js";
import { count } from "./src/fp/count.js";
import { difference } from "./src/fp/difference.js";
import { chunk } from "./src/fp/chunk.js";

const arr1 = [1,2,6,{a: 12, b: {c: "Foo"}}, 14, 14, 14, {a: 18, b: {c: "Foo"}}];

const arr2 = [
    [47,25,1,{ a: "Foo2", b: "Bar1", c: { d: "ct", e: { f: 9 }, g: 1 } }, [[800], [900, [1200, [622, 22], 555]]] ],
    [47,29,2,{ a: "Foo1", b: "Bar2", c: { d: "bg", e: { f: 5 }, g: 4 } }, [[801], [901, [1201, [650, 25], 560]]] ],
    [47,25,3,{ a: "Foo3", b: "Bar3", c: { d: "ce", e: { f: 7 }, g: 3 } }, [[802], [902, [1202, [622, 22], 570]]] ],
    [47,38,4,{ a: "Foo4", b: "Bar4", c: { d: "zd", e: { f: 1 }, g: 4 } }, [[803], [903, [1203, [650, 25], 580]]] ],
];

const arr3 = [
    [71,90,20,11,12],
    [72,91,50,30,13],
    [73,92,35,40,14],
    [74,93,86,50,15],
    [75,94,49,60,16],
]

/* (1) */
// const ex1 = Arr(arr1).count(14);
// console.log(ex1);

/* (2) */
// const ex2 = Arr(arr1);
// console.log(ex2.value);
// ex2.clear();
// console.log(ex2.value);
// console.log(arr1)

/* (3) */
// console.log(Arr(arr2).column(1).value);
// console.log(Arr(arr2).column(3, "a").value);
// console.log(Arr(arr2).column(3, "c.d").value);
// console.log(Arr(arr2).column(3, "c.e.f").value);
// console.log(Arr(arr2).column(item => item[3].c.e.f * 10).value);

// console.log(Arr(arr2).column(1).concat([77,77,77]));

// const col1 = Arr(arr2).column(1).entries();
// console.log(col1.next());
// console.log(col1.next());

// console.log(Arr(arr2).column(1).value);
// console.log(Arr(arr2).column(1).has(28));
// console.log(Arr(arr2).row(0).hasNullish());


// console.log(Arr(arr2).map(item => item[3].c.e.f));
// console.log(Arr(arr2).sumOf(item => item[3].c.e.f));
// console.log(Arr(arr2).sumOf("2"));
// console.log(Arr(arr2).sumOf("3.c.e.f"));
// console.log(Arr(arr2).productOf("3.c.e.f"));

// console.log(Arr(arr1).xunshift(70,80).value);
// console.log(Arr(arr1).xunshift(70,80).value);
// console.log(Arr(arr2).column(0).union([800], [900]).value);
// console.log(Arr(arr2).column(0).union([[800], [900, [1200, [622, 22], 555]]]).value);
// console.log(Arr(arr2).column(2).findAllOccurences(3).value);

// console.log(Arr(arr2).column(2).max());
// console.log(Arr(arr2).maxOf("2"));
// console.log(Arr(arr2).minOf("2"));
// console.log(Arr(arr2).maxOf("3.c.e.f"));
// console.log(Arr(arr2).maxBy(item => item[2]));
// console.log(Arr(arr2).maxBy("2"));
// console.log(Arr(arr2).maxBy("3.c.e.f"));
// console.log(Arr(arr2).minBy("3.c.e.f"));

// console.log(Arr(arr2).uniqueBy(item => item[1]).value);
// console.log(Arr(arr2).uniqueBy("3.c.g").value);
// console.log(Arr(arr2).column(1).types().isSingle());

// const vectorA = Arr(arr2).column(0);
// const vectorB = Arr(arr2).column(2);
// const vectorC = Arr(arr2).column(3, "c.e.f");
// const vectorC2 = Arr(arr2).column(item => item[3].c.e.f);
// console.log(vectorA.dotProduct(vectorB.value));
// console.log(vectorC2.xpop().crossProduct(vectorA.xpop().value));

// console.log(Arr([]).range(15).value);
// console.log(Arr([]).range(45,53).value);

// const r = Arr([1,2,3]).readOnly();
// r.push(50) // -> Error

// const arr = Arr(arr2).swapByIndexes(1,3).value;
// console.log(arr);

// const arr = Arr(arr2).swapByValues(arr2[0],arr2[1]).value;
// console.log(arr);

// const arr = Arr(arr2).delete(1,2);
// console.log(arr);

// const arr = Arr(arr2).clearByIndexes(0,1);
// console.log(arr);

// const arr = Arr(arr2).column(2);
// console.log(arr.value);
// console.log(arr.findAllOccurences(3));
// arr.clearByValues(2,4)
// arr.deleteByValues(2,4)
// arr.clearByIndexes(2,4)
// arr.deleteByIndexes(2,4);

// console.log(arr.value);
// arr.splice(0,1)
// console.log(arr.value);

// console.log(arr.value);
// console.log(arr.xsplice(0,1).value);
// console.log(Arr(arr2).column(2).value);
// console.log(arr.value);
// console.log(arr.xsplice(1,1).value);
// console.log(arr.splice(1,1).value);

// console.log(Arr(arr1).replaceByValue(14, 14004).value);
// console.log(Arr(arr1).replaceByValue(14, 14004, false).value);
// console.log(Arr(arr1).replaceByValues([14, 1], [140, 100]).value);
// console.log(Arr(arr1).replaceByIndex(3, 7000).value);
// console.log(Arr([ 'bg', 'ce', 'ct', 'zd' ]).reverse().value);
// console.log(Arr(arr2).column(3, "c.d").sort({ order: "descending" }).value);
// console.log(Arr([]).range(5000).shuffle().sort({ order: "descending", sortType: "HeapSort" }).slice(0,11).value);
// console.log(Arr([7,2,5,9,0,14,11]).sort({}).value);
// console.log(Arr(arr2).sort({ sortBy: item => item[3].c.e.f }).value);
// console.log(Arr([1,2,3]).interpolate([2, undefined, 6]));

const array1 = [1,2,3,4,5,6]
const array2 = [1,2,9,7,8,6]

console.log(difference(array1, array2));

// console.log(Arr(array1).difference(array2).value)
const array3 = [1,2,9,7,8,6,2,6,2,7]
console.log(count(2, array3))
console.log(chunk(3, array3));
